File: packages/sdk-py/aegis/webhook_events.py
"""Kind-discriminated webhook event union — Python mirror of
``packages/sdk-ts/src/webhook-events.ts``.

Webhook deliveries arrive at customer endpoints as JSON bodies with at
least ``event`` (the type name from the WEBHOOK_EVENT catalog) plus a
``data`` payload whose shape depends on the event. Without a typed
union, customers must access ``event["data"]`` blind and lose all
narrowing on payload shape.

This module ships:

- A :class:`WebhookEnvelope` :class:`TypedDict` union — each variant
  declares its ``event`` field as a :class:`Literal` so mypy narrows
  on ``if event["event"] == "..."``.
- :func:`interpret_webhook_event` — runtime validator that accepts a
  raw parsed-JSON dict and returns the union-typed envelope, raising
  :class:`WebhookEventParseError` on unknown events.
- :func:`is_webhook_envelope` — :class:`TypeGuard` variant for
  silent-skip patterns.
- A module-import-time catalog-coverage assertion that locks the
  Literal type against ``WEBHOOK_EVENT`` from
  ``_shared_constants_generated.py``: if the canonical catalog grows
  a new entry and this file isn't updated, import fails loudly
  rather than silently dropping events.

Customer pattern (after :func:`verify_webhook_signature` +
:func:`assert_not_replay`)::

    event = interpret_webhook_event(json.loads(verified_body))
    if event["event"] == "aegis.agent.trust_score_changed":
        # mypy narrows event["data"] to AgentTrustScoreChangedPayload
        score = event["data"]["score"]
        prev = event["data"]["previousScore"]
        ...
    elif event["event"] == "aegis.agent.policy_expired":
        # event["data"] is AgentPolicyExpiredPayload
        await revoke_downstream_session(event["data"]["agentId"])
        ...

Forward-compat: when a NEW event type ships in WEBHOOK_EVENT, the
import-time assertion below fails — forces an update to the union AND
the :func:`interpret_webhook_event` switch in the same release. Customer
code using ``if/elif`` on ``event["event"]`` also benefits from
:func:`typing.assert_never` exhaustiveness checks under mypy strict.

Payload schemas mirror the TS choice:

- For events with a known emitter
  (``aegis.agent.policy_expired``, ``aegis.agent.trust_score_changed``),
  we ship concrete payload :class:`TypedDict` shapes sourced from the
  API emitter code.
- For events declared in the catalog but not yet emitted
  (``anomaly_detected``, ``flagged_by_relying_party``, ``revoked``),
  payload is ``dict[str, Any]`` — concrete schemas land WITH the
  emitter, per CLAUDE.md docs rule "docs reflect code, not aspiration".
"""

from __future__ import annotations

from typing import Any, Literal, NotRequired, TypedDict, TypeGuard

from ._shared_constants_generated import WEBHOOK_EVENT

# ── Event-name literal (locked against WEBHOOK_EVENT catalog) ─────────
# Exact mirror of `WebhookEvent` in @aegis/types. If the catalog grows
# a new entry, the import-time assertion at the bottom of this module
# fails with a clear error pointing at the missing variant — the same
# discipline as the TS ``_ExhaustivenessGate``.
WebhookEventName = Literal[
    "aegis.agent.trust_score_changed",
    "aegis.agent.anomaly_detected",
    "aegis.agent.policy_expired",
    "aegis.agent.flagged_by_relying_party",
    "aegis.agent.revoked",
]


# ── Payload shapes ────────────────────────────────────────────────────


class AgentTrustScoreChangedPayload(TypedDict):
    """Emitted when an agent's trust score crosses a band boundary.

    Source: ``apps/api/src/modules/bate/bate.worker.ts:249``.
    """

    agentId: str
    score: int
    previousScore: int
    band: str
    previousBand: str
    weightsVersion: str
    contributors: dict[str, int]


class AgentPolicyExpiredPayload(TypedDict):
    """Emitted when an agent's policy passes its ``expiresAt`` timestamp.

    Source: ``apps/api/src/modules/policy/policy.expiry.worker.ts:144``.
    """

    policyId: str
    agentId: str
    expiredAt: str  # ISO-8601 timestamp the policy was scheduled to expire
    sweptAt: str  # ISO-8601 timestamp the expiry sweep ran


# Placeholder payload types for events with no emitter yet. Concrete
# schemas land WITH the emitter — mirrors TS's `Record<string, unknown>`.
AgentAnomalyDetectedPayload = dict[str, Any]
AgentFlaggedByRelyingPartyPayload = dict[str, Any]
AgentRevokedPayload = dict[str, Any]


# ── Envelope variants ─────────────────────────────────────────────────


class AgentTrustScoreChangedEvent(TypedDict):
    """Webhook envelope for ``aegis.agent.trust_score_changed``."""

    event: Literal["aegis.agent.trust_score_changed"]
    data: AgentTrustScoreChangedPayload
    subscriptionId: NotRequired[str]
    deliveryId: NotRequired[str]
    occurredAt: NotRequired[str]


class AgentAnomalyDetectedEvent(TypedDict):
    """Webhook envelope for ``aegis.agent.anomaly_detected``."""

    event: Literal["aegis.agent.anomaly_detected"]
    data: AgentAnomalyDetectedPayload
    subscriptionId: NotRequired[str]
    deliveryId: NotRequired[str]
    occurredAt: NotRequired[str]


class AgentPolicyExpiredEvent(TypedDict):
    """Webhook envelope for ``aegis.agent.policy_expired``."""

    event: Literal["aegis.agent.policy_expired"]
    data: AgentPolicyExpiredPayload
    subscriptionId: NotRequired[str]
    deliveryId: NotRequired[str]
    occurredAt: NotRequired[str]


class AgentFlaggedByRelyingPartyEvent(TypedDict):
    """Webhook envelope for ``aegis.agent.flagged_by_relying_party``."""

    event: Literal["aegis.agent.flagged_by_relying_party"]
    data: AgentFlaggedByRelyingPartyPayload
    subscriptionId: NotRequired[str]
    deliveryId: NotRequired[str]
    occurredAt: NotRequired[str]


class AgentRevokedEvent(TypedDict):
    """Webhook envelope for ``aegis.agent.revoked``."""

    event: Literal["aegis.agent.revoked"]
    data: AgentRevokedPayload
    subscriptionId: NotRequired[str]
    deliveryId: NotRequired[str]
    occurredAt: NotRequired[str]


# The full kind-discriminated webhook event union. Match/check on
# ``envelope["event"]`` to narrow to the concrete envelope variant.
WebhookEnvelope = (
    AgentTrustScoreChangedEvent
    | AgentAnomalyDetectedEvent
    | AgentPolicyExpiredEvent
    | AgentFlaggedByRelyingPartyEvent
    | AgentRevokedEvent
)


# ── Errors ────────────────────────────────────────────────────────────


class WebhookEventParseError(ValueError):
    """Raised when an envelope doesn't match any known event.

    Inherits from :class:`ValueError` (not :class:`AegisError`) because
    parse failures are caller-input errors, not API errors — caller
    likely deserialized a non-envelope JSON body or is running an
    SDK older than the API. The accompanying guidance message points
    at the most common cause.

    Attributes:
        raw_event_name: The raw ``event`` value that failed to map.
            ``None`` if the envelope was missing the discriminator
            entirely or was not a dict.
    """

    def __init__(self, message: str, *, raw_event_name: object | None) -> None:
        super().__init__(message)
        self.raw_event_name = raw_event_name


# ── Import-time exhaustiveness gate ───────────────────────────────────
#
# Lock the WebhookEventName Literal against the WEBHOOK_EVENT catalog.
# If a new event is added to ``_shared_constants_generated.py`` (which
# is itself regenerated from ``packages/types/src/constants.ts``) but
# this module isn't updated, this assertion fails at import time —
# loud enough that any consumer of `aegis` notices, and the
# cross-language parity gate catches it earlier in CI.
#
# The set of catalog values must EXACTLY equal the set of strings
# expressible as ``WebhookEventName``. We can't introspect a
# :class:`Literal` at runtime cleanly, so we hard-code the expected set
# and assert both directions: (1) every catalog value is one of ours,
# (2) every one of ours is in the catalog.

_KNOWN_EVENT_NAMES: frozenset[str] = frozenset(
    {
        "aegis.agent.trust_score_changed",
        "aegis.agent.anomaly_detected",
        "aegis.agent.policy_expired",
        "aegis.agent.flagged_by_relying_party",
        "aegis.agent.revoked",
    }
)

_catalog_values: frozenset[str] = frozenset(WEBHOOK_EVENT.values())

_missing_from_module: frozenset[str] = _catalog_values - _KNOWN_EVENT_NAMES
if _missing_from_module:
    raise RuntimeError(
        f"webhook_events.py is out of date: WEBHOOK_EVENT catalog has "
        f"new entries not yet declared as WebhookEnvelope variants: "
        f"{sorted(_missing_from_module)}. Add the variants AND update "
        f"_KNOWN_EVENT_NAMES AND update interpret_webhook_event."
    )

_extra_in_module: frozenset[str] = _KNOWN_EVENT_NAMES - _catalog_values
if _extra_in_module:
    raise RuntimeError(
        f"webhook_events.py declares variants that aren't in WEBHOOK_EVENT "
        f"catalog: {sorted(_extra_in_module)}. Either remove the variants "
        f"or update _shared_constants_generated.py / @aegis/types."
    )


# ── interpret helper ──────────────────────────────────────────────────


def interpret_webhook_event(raw: object) -> WebhookEnvelope:
    """Narrow a raw envelope (typically ``json.loads(verified_body)``) into
    the typed :class:`WebhookEnvelope` union.

    Resolution:

    1. Validate the envelope is a :class:`dict`.
    2. Validate ``event`` is a string and one of the known catalog
       entries. Unknown → :class:`WebhookEventParseError` with a
       guidance message about SDK upgrade.
    3. Cast the raw dict through to the narrowed union type. No
       payload shape validation — that is the customer's responsibility
       (use a runtime guard, pydantic schema, etc.). Validating here
       would require shipping schemas for every payload, which we don't
       have for events with no emitter (matches TS choice).

    Args:
        raw: A parsed-JSON value — typically ``json.loads(verified_body)``.

    Returns:
        The same dict, narrowed to :class:`WebhookEnvelope`.

    Raises:
        WebhookEventParseError: ``raw`` is not a dict, ``event`` is
            missing/not-a-string, or ``event`` is not one of the
            known catalog values.
    """
    if not isinstance(raw, dict):
        raise WebhookEventParseError(
            f"webhook envelope must be a dict, got {type(raw).__name__}",
            raw_event_name=None,
        )

    event_name = raw.get("event")
    if not isinstance(event_name, str):
        raise WebhookEventParseError(
            f"webhook envelope missing string 'event' field, got "
            f"{type(event_name).__name__}",
            raw_event_name=event_name,
        )

    if event_name not in _KNOWN_EVENT_NAMES:
        raise WebhookEventParseError(
            f"webhook envelope has unknown event name: {event_name!r}. "
            f"SDK may be older than the API — consider upgrading the "
            f"`aegis` package.",
            raw_event_name=event_name,
        )

    # Trust the post-signature-verify envelope shape. The import-time
    # catalog-coverage assertion above has already proved every catalog
    # value is a valid variant; the runtime cast is safe.
    return raw  # type: ignore[return-value]


def is_webhook_envelope(raw: object) -> TypeGuard[WebhookEnvelope]:
    """Type-guard variant of :func:`interpret_webhook_event`.

    Returns ``True`` and narrows ``raw`` when the envelope is a known
    webhook event; returns ``False`` otherwise without raising. Use
    when you want to silently skip unknown events (e.g. during SDK-
    upgrade transitions when the API may emit events the SDK doesn't
    recognize yet).
    """
    try:
        interpret_webhook_event(raw)
    except WebhookEventParseError:
        return False
    else:
        return True


__all__ = [
    "AgentAnomalyDetectedEvent",
    "AgentAnomalyDetectedPayload",
    "AgentFlaggedByRelyingPartyEvent",
    "AgentFlaggedByRelyingPartyPayload",
    "AgentPolicyExpiredEvent",
    "AgentPolicyExpiredPayload",
    "AgentRevokedEvent",
    "AgentRevokedPayload",
    "AgentTrustScoreChangedEvent",
    "AgentTrustScoreChangedPayload",
    "WebhookEnvelope",
    "WebhookEventName",
    "WebhookEventParseError",
    "interpret_webhook_event",
    "is_webhook_envelope",
]

File: packages/sdk-ts/src/webhook-events.ts
// Kind-discriminated webhook event union.
//
// Webhook deliveries arrive at customer endpoints as JSON bodies with
// at least `event` (the type name from WEBHOOK_EVENT catalog) plus a
// `data` payload whose shape depends on the event. Without a typed
// union, customers must cast `event.event` to `any` and lose all
// compile-time guarantees on payload narrowing.
//
// This module ships the union + an `interpretWebhookEvent` helper:
//
//   import { verifyWebhookSignature, interpretWebhookEvent } from '@aegis/sdk';
//
//   // verify first — never narrow untrusted bodies
//   await verifyWebhookSignature({ payload: raw, signature, secret });
//   const event = interpretWebhookEvent(JSON.parse(raw));
//   switch (event.event) {
//     case 'aegis.agent.trust_score_changed':
//       // event.data is narrowed to AgentTrustScoreChangedPayload here
//       console.log(event.data.previousBand, '→', event.data.band);
//       break;
//     // ... tsc fails if any catalog event is missing here ...
//   }
//
// Forward-compat: when a NEW event type ships in WEBHOOK_EVENT, the
// `_ExhaustivenessGate` below stops compiling — tsc forces an update
// to the union AND the interpret() switch. The customer code that
// uses `switch (event.event)` also stops compiling on the customer
// side because of the same `never` narrowing, so callers can't ship
// a release that silently drops new event types.
//
// Payload schemas:
//   - For events with a known emitter (aegis.agent.policy_expired,
//     aegis.agent.trust_score_changed), we ship concrete payload
//     interfaces sourced from the API emitter code.
//   - For events declared in the catalog but not yet emitted
//     (anomaly_detected, flagged_by_relying_party, revoked), we ship
//     `Record<string, unknown>` so customers don't code against a
//     guessed shape. Concrete schemas land WITH the emitter, not
//     before it, per CLAUDE.md docs rule "docs reflect code, not
//     aspiration".

import { WEBHOOK_EVENT, type WebhookEvent } from '@aegis/types';

import type { TrustBand } from './types.js';

// ── Payload shapes ──────────────────────────────────────────────

/**
 * Emitted when an agent's trust score crosses a band boundary.
 * Source: `apps/api/src/modules/bate/bate.worker.ts:249`.
 */
export interface AgentTrustScoreChangedPayload {
  agentId: string;
  /** New trust score (0..1000). */
  score: number;
  /** Previous trust score. */
  previousScore: number;
  /** New trust band — one of PLATINUM/VERIFIED/WATCH/FLAGGED. */
  band: TrustBand;
  /** Previous trust band. */
  previousBand: TrustBand;
  /** BATE algorithm weights version that produced this score. */
  weightsVersion: string;
  /** Per-signal contribution breakdown (signal-type → delta). */
  contributors: Record<string, number>;
}

/**
 * Emitted when an agent's policy passes its `expiresAt` timestamp.
 * Source: `apps/api/src/modules/policy/policy.expiry.worker.ts:144`.
 */
export interface AgentPolicyExpiredPayload {
  policyId: string;
  agentId: string;
  /** ISO-8601 timestamp the policy was scheduled to expire. */
  expiredAt: string;
  /** ISO-8601 timestamp the expiry sweep ran. */
  sweptAt: string;
}

/**
 * Reserved for the BATE anomaly detector. Payload schema lands with
 * the emitter — until then, customers should narrow defensively.
 */
export type AgentAnomalyDetectedPayload = Record<string, unknown>;

/**
 * Reserved for relying-party-initiated agent flagging. Payload schema
 * lands with the emitter.
 */
export type AgentFlaggedByRelyingPartyPayload = Record<string, unknown>;

/**
 * Reserved for manual agent revocation events. Payload schema lands
 * with the emitter.
 */
export type AgentRevokedPayload = Record<string, unknown>;

// ── Envelope union ──────────────────────────────────────────────

interface WebhookEnvelopeBase {
  /** Webhook event name from the WEBHOOK_EVENT catalog. */
  event: WebhookEvent;
  /** Subscription ID this delivery targets. */
  subscriptionId?: string;
  /** Delivery ID — matches `X-AEGIS-Delivery-Id` header. */
  deliveryId?: string;
  /** ISO-8601 timestamp the event occurred (independent of delivery time). */
  occurredAt?: string;
}

export interface AgentTrustScoreChangedEvent extends WebhookEnvelopeBase {
  event: typeof WEBHOOK_EVENT.AGENT_TRUST_SCORE_CHANGED;
  data: AgentTrustScoreChangedPayload;
}
export interface AgentAnomalyDetectedEvent extends WebhookEnvelopeBase {
  event: typeof WEBHOOK_EVENT.AGENT_ANOMALY_DETECTED;
  data: AgentAnomalyDetectedPayload;
}
export interface AgentPolicyExpiredEvent extends WebhookEnvelopeBase {
  event: typeof WEBHOOK_EVENT.AGENT_POLICY_EXPIRED;
  data: AgentPolicyExpiredPayload;
}
export interface AgentFlaggedByRelyingPartyEvent extends WebhookEnvelopeBase {
  event: typeof WEBHOOK_EVENT.AGENT_FLAGGED_BY_RELYING_PARTY;
  data: AgentFlaggedByRelyingPartyPayload;
}
export interface AgentRevokedEvent extends WebhookEnvelopeBase {
  event: typeof WEBHOOK_EVENT.AGENT_REVOKED;
  data: AgentRevokedPayload;
}

/**
 * The full kind-discriminated webhook event union. Switch on
 * `envelope.event` to narrow to the concrete payload type.
 */
export type WebhookEnvelope =
  | AgentTrustScoreChangedEvent
  | AgentAnomalyDetectedEvent
  | AgentPolicyExpiredEvent
  | AgentFlaggedByRelyingPartyEvent
  | AgentRevokedEvent;

// ── Exhaustiveness gate ─────────────────────────────────────────

// type-rationale: this Record forces tsc to verify EVERY value in
// the WEBHOOK_EVENT catalog appears as the `event` field of at least
// one variant in the WebhookEnvelope union. If a new event is added
// to the catalog (`@aegis/types`) but not to the union here, the
// type evaluates `never` for that key and the const declaration
// fails to compile with a clear error pointing at the missing
// variant. Mirrors the verify-outcome union's compile-time gate.
type _ExhaustivenessGate = {
  [K in WebhookEvent]: Extract<WebhookEnvelope, { event: K }> extends never
    ? `MISSING WebhookEnvelope variant for event: ${K}`
    : true;
};
const _exhaustivenessProof: _ExhaustivenessGate = {
  [WEBHOOK_EVENT.AGENT_TRUST_SCORE_CHANGED]: true,
  [WEBHOOK_EVENT.AGENT_ANOMALY_DETECTED]: true,
  [WEBHOOK_EVENT.AGENT_POLICY_EXPIRED]: true,
  [WEBHOOK_EVENT.AGENT_FLAGGED_BY_RELYING_PARTY]: true,
  [WEBHOOK_EVENT.AGENT_REVOKED]: true,
};
// Reference the proof so unused-symbol lints don't strip it. The
// `void` keeps it side-effect-free in the bundle.
void _exhaustivenessProof;

// ── Interpret helper ────────────────────────────────────────────

/**
 * Custom error for envelopes that don't match any known event. Thrown
 * by `interpretWebhookEvent` so callers can distinguish a parsing
 * failure from an unsupported event (e.g. a future event whose SDK
 * release hasn't shipped yet).
 */
export class WebhookEventParseError extends Error {
  override readonly name = 'WebhookEventParseError';
  constructor(
    message: string,
    /** The raw `event` value that failed to map, if present. */
    public readonly rawEventName: unknown,
  ) {
    super(message);
  }
}

/**
 * Narrow a raw envelope (typically `JSON.parse(verifiedBody)`) into
 * the typed `WebhookEnvelope` union. Throws `WebhookEventParseError`
 * if the envelope is missing the discriminator or carries an
 * unrecognized event name.
 *
 * Resolution:
 *   1. Validate the envelope is an object with a string `event` field.
 *   2. Validate `event` is one of the known catalog entries (the
 *      switch is exhaustive via the `_never` narrowing at the bottom
 *      — adding a catalog entry without a case here is a tsc error).
 *   3. Pass the raw envelope through with its narrowed type.
 *
 * Note: this helper does NOT validate the `data` payload shape — that
 * is the customer's responsibility (Zod schema, runtime guard, etc.).
 * Validating here would require shipping Zod schemas for every payload,
 * which we don't yet have for events with no emitter. For now: trust
 * the post-signature-verify body and validate at the use site.
 */
export function interpretWebhookEvent(raw: unknown): WebhookEnvelope {
  if (typeof raw !== 'object' || raw === null) {
    throw new WebhookEventParseError(
      'webhook envelope must be a non-null object',
      raw,
    );
  }
  const event = (raw as { event?: unknown }).event;
  if (typeof event !== 'string') {
    throw new WebhookEventParseError(
      `webhook envelope missing string 'event' field, got ${typeof event}`,
      event,
    );
  }
  switch (event) {
    case WEBHOOK_EVENT.AGENT_TRUST_SCORE_CHANGED:
    case WEBHOOK_EVENT.AGENT_ANOMALY_DETECTED:
    case WEBHOOK_EVENT.AGENT_POLICY_EXPIRED:
    case WEBHOOK_EVENT.AGENT_FLAGGED_BY_RELYING_PARTY:
    case WEBHOOK_EVENT.AGENT_REVOKED:
      // Trust the post-signature-verify envelope shape. The static
      // type system has already proved (via _ExhaustivenessGate
      // above) that every catalog value is covered here; the runtime
      // cast is safe.
      return raw as WebhookEnvelope;
    default: {
      throw new WebhookEventParseError(
        `webhook envelope has unknown event name: ${JSON.stringify(event)}. ` +
          `SDK may be older than the API — consider upgrading @aegis/sdk.`,
        event,
      );
    }
  }
}

/**
 * Type guard variant of `interpretWebhookEvent`. Returns `true` and
 * narrows the input when the envelope is a known webhook event;
 * returns `false` otherwise without throwing. Use when you want to
 * silently skip unknown events (e.g. during SDK-upgrade transitions).
 */
export function isWebhookEnvelope(raw: unknown): raw is WebhookEnvelope {
  try {
    interpretWebhookEvent(raw);
    return true;
  } catch {
    return false;
  }
}
